### ISSUE TITLE
runtime error while running mace with opencl

### ISSUE THREAD

**Issue by reasonsolo** [XiaoMi/mace#232]
I'm trying to run mace on my device(mali GPU).

### System information
-  android 7.12 on RK3399
-  ndk-17c with clang toolset
-  MACE version: v0.9.0-282-g9b78e12-20181029

By ` python tools/converter.py run --config=mace-models/mobilenet-v2/profile.yaml   --validate
`, I got following output.

```
WARNING: linker: /system/lib64/libmace.so: unused DT entry: type 0xf arg 0x13489f
I mace.cc:629 Create MaceEngine from model pb
I mace.cc:387 Creating MaceEngine, MACE version: v0.9.0-282-g9b78e12-20181029
I mace.cc:483 Loading Model Data
I mace.cc:413 Initializing MaceEngine
W tensor.h:286 mobilenet_v2/Flatten/flatten/strided_slice:0: Resize buffer from size 65 to 66
W tensor.h:286 mobilenet_v2/Flatten/flatten/Reshape/shape:0: Resize buffer from size 66 to 68
F tensor.h:226 Check failed: has_opencl_image()  do not have image
Aborted 
```

Appraciate it if any one could give me a opencl enabled mace tutorial.

my model yaml
```
library_name: profile-v1031
target_abis: [arm64-v8a]
target_socs: [rk3399]
model_graph_format: file
model_data_format: file
models:
  profile_v1031:
    platform: tensorflow
    model_file_path: /tmp/profile_v1031.pb
    model_sha256_checksum: f63570f2d76fbbb99621fd24413451313b62874d9053b72583962c7d7a4be71e
    subgraphs:
      - input_tensors:
          - input_image
        input_shapes:
          - 1,96,96,3
        output_tensors:
          - gender
          - age
        output_shapes:
          - 1,2
          - 1,101
    runtime: gpu
    limit_opencl_kernel_time: 0
    nnlib_graph_mode: 0
    obfuscate: 1
    winograd: 0
    cl_mem_type: image
```

**Comment by reasonsolo** [comment]
my model yaml

```
library_name: profile-v1031
target_abis: [arm64-v8a]
target_socs: [rk3399]
model_graph_format: file
model_data_format: file
models:
  profile_v1031:
    platform: tensorflow
    model_file_path: /tmp/profile_v1031.pb
    model_sha256_checksum: f63570f2d76fbbb99621fd24413451313b62874d9053b72583962c7d7a4be71e
    subgraphs:
      - input_tensors:
          - input_image
        input_shapes:
          - 1,96,96,3
        output_tensors:
          - gender
          - age
        output_shapes:
          - 1,2
          - 1,101
    runtime: gpu
    limit_opencl_kernel_time: 0
    nnlib_graph_mode: 0
    obfuscate: 1
    winograd: 0
    cl_mem_type: image

```

**Comment by TonyMou** [comment]
@reasonsolo , it looks like some op in your model is not supported by gpu. I'll take a look at your pb file.

**Comment by reasonsolo** [comment]
```
I net.cc:83 Operator ab651090 type: Conv2D future 0
I net.cc:138 op done, has shape: [1, 3, 3, 1280]
I net.cc:83 Operator 8cd6923e type: Pooling future 0
I net.cc:138 op done, has shape: [1, 1, 1, 1280]
I net.cc:83 Operator 42afec5f type: Conv2D future 0
I net.cc:138 op done, has shape: [1, 1, 1, 512]
I net.cc:83 Operator c7efede0 type: StridedSlice future 0
W tensor.h:286 a67bf373: Resize buffer from size 65 to 66
I net.cc:138 op done, has shape: []
I net.cc:83 Operator 2ac076a8 type: Stack future 0
W tensor.h:286 9d3dcab9: Resize buffer from size 66 to 68
I net.cc:138 op done, has shape: [2]
I net.cc:83 Operator 9d6a9690 type: Reshape future 0
I net.cc:138 op done, has shape: [1, 512]
I net.cc:83 Operator b8c7d6fc type: MatMul future 0
F tensor.h:226 Check failed: has_opencl_image()  do not have image, type N4mace6BufferE
```
So I guess the op not supported by gpu is reshape? I checked through kernels code and no reshape.cc is found in kernels/opencl :(

**Comment by TonyMou** [comment]
The problem is " Matmul " Op in "logits/gender" and "logits/age" does not support weights data as input. You can use FullyConnect Op instead.

**Comment by reasonsolo** [comment]
Thank you. I will try to remove last reshape/matmul ops while running inference.

But is it possible to use a matmul_buffer cl kernel to do the trick? Or why mace chose to run matmul for unsupported data type?

**Comment by liyinhgqw** [comment]
We will work on it and see what is the proper way

**Comment by reasonsolo** [comment]
In my case,  the inputs of matmul ops A is cl image while B is cl buffer. Is this a cl mem allocating problem?
```
I matmul.h:107 A is image shape [1, 512]
I matmul.h:116 B is buffer, shape [512, 2]
```

**Comment by liyinhgqw** [comment]
No, it is because we do not support 'MatMul' but support 'FC' for GPU. We'll consider implementing this.

For now, you can try FC.

**Comment by wayen820** [comment]
@liyinhgqw how can i use FC when i use tensorflow pb ,I see tools/convert_tool/tensorflow_converter.py,there do not have any tensorflow ops convert to mace FC op.Thanks
